src/iop/lmmse_demosaic.c

```diff
@@ -0,0 +1,602 @@
+  float *qix[5];
+  float *buffer = dt_alloc_align_float(grp_height * grp_width * 5);
+
+  if(!buffer)
+  {
+    dt_free_align(buffer);
+    dt_control_log(_("[lmmse_demosaic] can't allocate buffer"));
+    fprintf(stderr, "[lmmse_demosaic] can't allocate buffer\n");
+    return;
+  }
+
+  qix[0] = buffer;
+  for(int i = 1; i < 5; i++)
+  {
+    qix[i] = qix[i - 1] + grp_height * grp_width;
+  }
+
+  memset(buffer, 0, sizeof(float) * grp_height * grp_width * 5);
+
+  const float scaler = fmaxf(piece->pipe->dsc.processed_maximum[0], fmaxf(piece->pipe->dsc.processed_maximum[1], piece->pipe->dsc.processed_maximum[2]));
+  const float revscaler = 1.0f / scaler;
+
+#ifdef _OPENMP
+    #pragma omp parallel private(rix)
+#endif
+  {
+#ifdef _OPENMP
+    #pragma omp for
+#endif
+    for(int rrr = BORDER_AROUND; rrr < grp_height - BORDER_AROUND; rrr++)
+    {
+      for(int ccc = BORDER_AROUND, row = rrr - BORDER_AROUND; ccc < grp_width - BORDER_AROUND; ccc++)
+      {
+        float *rix0 = qix[4] + rrr * grp_width + ccc;
+        rix0[0] = calc_gamma(revscaler * in[row * width + ccc - BORDER_AROUND], gamma_in);
+      }
+    }
+
+    // G-R(B)
+#ifdef _OPENMP
+    #pragma omp for schedule(dynamic,16)
+#endif
+    for(int rr = 2; rr < grp_height - 2; rr++)
+    {
+      // G-R(B) at R(B) location
+      for(int cc = 2 + (FC(rr, 2, filters) & 1); cc < grp_width - 2; cc += 2)
+      {
+        rix[4] = qix[4] + rr * grp_width + cc;
+        float v0 = 0.0625f * (rix[4][-w1 - 1] + rix[4][-w1 + 1] + rix[4][w1 - 1] + rix[4][w1 + 1]) + 0.25f * rix[4][0];
+        // horizontal
+        rix[0] = qix[0] + rr * grp_width + cc;
+        rix[0][0] = -0.25f * (rix[4][ -2] + rix[4][ 2]) + 0.5f * (rix[4][ -1] + rix[4][0] + rix[4][ 1]);
+        float Y = v0 + 0.5f * rix[0][0];
+
+        rix[0][0] = (rix[4][0] > 1.75f * Y) ? median3f(rix[0][0], rix[4][ -1], rix[4][ 1]) : limf(rix[0][0], 0.0f, 1.0f);
+        rix[0][0] -= rix[4][0];
+        // vertical
+        rix[1] = qix[1] + rr * grp_width + cc;
+        rix[1][0] = -0.25f * (rix[4][-w2] + rix[4][w2]) + 0.5f * (rix[4][-w1] + rix[4][0] + rix[4][w1]);
+        Y = v0 + 0.5f * rix[1][0];
+
+        rix[1][0] = (rix[4][0] > 1.75f * Y) ? median3f(rix[1][0], rix[4][-w1], rix[4][w1]) : limf(rix[1][0], 0.0f, 1.0f);
+        rix[1][0] -= rix[4][0];
+      }
+
+      // G-R(B) at G location
+      for(int ccc = 2 + (FC(rr, 3, filters) & 1); ccc < grp_width - 2; ccc += 2)
+      {
+        rix[0] = qix[0] + rr * grp_width + ccc;
+        rix[1] = qix[1] + rr * grp_width + ccc;
+        rix[4] = qix[4] + rr * grp_width + ccc;
+        rix[0][0] = 0.25f * (rix[4][ -2] + rix[4][ 2]) - 0.5f * (rix[4][ -1] + rix[4][0] + rix[4][ 1]);
+        rix[1][0] = 0.25f * (rix[4][-w2] + rix[4][w2]) - 0.5f * (rix[4][-w1] + rix[4][0] + rix[4][w1]);
+        rix[0][0] = limf(rix[0][0], -1.0f, 0.0f) + rix[4][0];
+        rix[1][0] = limf(rix[1][0], -1.0f, 0.0f) + rix[4][0];
+      }
+    }
+
+    // apply low pass filter on differential colors
```

Comment:
Be aware that this section makes heavy use of negative indices for arrays, which tools like `cppcheck` complain about. The code is valid though.

Reply:
ok. BTW after my fixing the currently still wrong refinement code (wrong input range) i would like to improve performance. From my understanding of commits in the rt codebase you seem to know the code pretty well. Interested in sort of collaboration on that? Via private mail or pixels.us ?

Reply:
Hi @jenshannoschwalm not nearly as well as Ingo, I've not dived into these kinds of algorithms before. But there could be a first time for everything. Let's use pixls for a chat.